x-pack/plugins/rule_registry/server/alert_data_client/alerts_client.ts

```diff
@@ -0,0 +1,215 @@
+  return source?.[RULE_ID] != null && source?.[OWNER] != null;
+};
+export interface ConstructorOptions {
+  logger: Logger;
+  authorization: PublicMethodsOf<AlertingAuthorization>;
+  auditLogger?: AuditLogger;
+  esClient: ElasticsearchClient;
+}
+
+export interface UpdateOptions<Params extends AlertTypeParams> {
+  id: string;
+  data: {
+    status: string;
+  };
+  index: string;
+}
+
+interface GetAlertParams {
+  id: string;
+  index?: string;
+}
+
+/**
+ * Provides apis to interact with alerts as data
+ * ensures the request is authorized to perform read / write actions
+ * on alerts as data.
+ */
+export class AlertsClient {
+  private readonly logger: Logger;
+  private readonly auditLogger?: AuditLogger;
+  private readonly authorization: PublicMethodsOf<AlertingAuthorization>;
+  private readonly esClient: ElasticsearchClient;
+
+  constructor({ auditLogger, authorization, logger, esClient }: ConstructorOptions) {
+    this.logger = logger;
+    this.authorization = authorization;
+    this.esClient = esClient;
+    this.auditLogger = auditLogger;
+  }
+
+  public async getAlertsIndex(featureIds: string[]) {
+    return this.authorization.getAugmentRuleTypesWithAuthorization(
+      featureIds.length !== 0 ? featureIds : ['apm', 'siem']
+    );
+  }
+
+  private async fetchAlert({ id, index }: GetAlertParams): Promise<AlertType> {
+    try {
+      const result = await this.esClient.search<ParsedTechnicalFields>({
+        // Context: Originally thought of always just searching `.alerts-*` but that could
+        // result in a big performance hit. If the client already knows which index the alert
+        // belongs to, passing in the index will speed things up
+        index: index ?? '.alerts-*',
+        body: { query: { term: { _id: id } } },
+      });
+
+      if (!isValidAlert(result.body.hits.hits[0]._source)) {
+        const errorMessage = `Unable to retrieve alert details for alert with id of "${id}".`;
+        this.logger.debug(errorMessage);
+        throw new Error(errorMessage);
+      }
+
+      return result.body.hits.hits[0]._source;
+    } catch (error) {
+      const errorMessage = `Unable to retrieve alert with id of "${id}".`;
+      this.logger.debug(errorMessage);
+      throw error;
+    }
+  }
+
+  public async get({ id, index }: GetAlertParams): Promise<ParsedTechnicalFields> {
+    try {
+      // first search for the alert by id, then use the alert info to check if user has access to it
+      const alert = await this.fetchAlert({
+        id,
+        index,
+      });
+
+      // this.authorization leverages the alerting plugin's authorization
+      // client exposed to us for reuse
+      await this.authorization.ensureAuthorized({
+        ruleTypeId: alert[RULE_ID],
+        consumer: alert[OWNER],
+        operation: ReadOperations.Get,
+        entity: AlertingAuthorizationEntity.Alert,
+      });
+
+      this.auditLogger?.log(
```

Comment:
**question** If I'm reading this correctly, the alerting authorization client is performing its own audit logging. It seems that we should either be leveraging that audit log record, or suppress that audit log in favor of your own. What do you think makes more sense?

This comment applies to all audit events in this file, but I'll just write it once for brevity

Reply:
The alerting authorization client logs audit entries related to user access to rule type rules/alerts like:

```
server    log   [13:29:46.389] [info][alerting_authorization_success][alerts][audit][plugins][security] elastic Authorized to get a "example.always-firing" rule for "alerts"
```

This audit log record is similar to the alerting framework alerts client audit log entries, which logs audit events specific to the document/saved object being accessed, where the audit entries look like this:

```
{
	"event": {
		"action": "rule_get",
		"category": ["database"],
		"type": ["access"],
		"outcome": "success"
	},
	"kibana": {
		"space_id": "default",
		"saved_object": {
			"type": "alert",
			"id": "b6ce1de0-d447-11eb-9d81-099e8c561060"
		}
	},
	"trace": {
		"id": "571e0631-41b8-4a77-ba46-7cec8fdf8718"
	},
	"ecs": {
		"version": "1.9.0"
	},
	"@timestamp": "2021-06-23T13:35:08.756-04:00",
	"message": "User has accessed rule [id=b6ce1de0-d447-11eb-9d81-099e8c561060]",
	"log": {
		"level": "INFO",
		"logger": "plugins.security.audit.ecs"
	},
	"process": {
		"pid": 51636
	}
}
```

Reply:
Ah I see, thanks for clarifying @ymao1!